src/graphnet/data/pipeline.py

```diff
@@ -0,0 +1,173 @@
+        return dataloader
+
+    def _combine_outputs(self, dataframes):
+        return reduce(lambda x, y: pd.merge(x, y, on = 'event_no'), dataframes)
+
+    def _get_outdir(self, database):
+        if self._outdir == None:
+            database_name = database.split('/')[-3]
+            outdir = database.split(database_name)[0] + database_name +  '/pipelines/' + self._pipeline_name
+        else:
+            outdir = self._outdir
+        return outdir
+
+    def _get_truth(self,database):
+        with sqlite3.connect(database) as con:
+            query = 'SELECT * FROM truth'
+            truth = pd.read_sql(query,con)
+        return truth
+
+    def _get_retro(self,database):
+        try:
+            with sqlite3.connect(database) as con:
+                query = 'SELECT * FROM %s'%self._retro_table_name
+                retro = pd.read_sql(query,con)
+            return retro
+        except:
+            print('%s table does not exist'%self._retro_table_name)
+            return
+
+
+    def _make_pipeline_database(self,outdir, df, original_database):
+        os.makedirs(outdir)
+        pipeline_database = outdir + '/%s.db'%self._pipeline_name
+        truth = self._get_truth(original_database)
+        retro = self._get_retro(original_database)
+
+        self._save_to_sql(df, 'reconstruction', pipeline_database)
+        self._save_to_sql(truth, 'truth', pipeline_database)
+        if isinstance(retro, pd.DataFrame):
+            self._save_to_sql(retro, self._retro_table_name, pipeline_database)    
+        return
+        
+
+    def _save_to_sql(self, df, table_name, pipeline_database):
+        print('SUBMITTING %s'%table_name)
+        self._create_table(pipeline_database,  table_name, df)
+        engine = sqlalchemy.create_engine('sqlite:///' + pipeline_database)
+        df.to_sql(table_name, con=engine, index=False, if_exists='append')
+        engine.dispose()
+        print('%s SUBMITTED'%table_name)
+        return        
+
+
+    def _run_sql_code(self, database: str, code: str):
+        conn = sqlite3.connect(database)
+        c = conn.cursor()
+        c.executescript(code)
+        c.close()
+
+
+    def _create_table(self, pipeline_database, table_name, df):
+        """Creates a table.
+        Args:
+            pipeline_database (str): path to the pipeline database
+            df (str): pandas.DataFrame of combined predictions
+        """
+        query_columns = list()
+        for column in df.columns:
+            if column == 'event_no':
+                type_ = 'INTEGER PRIMARY KEY NOT NULL'
+            else:
+                type_ = 'FLOAT'
+            query_columns.append(f"{column} {type_}")
+        query_columns = ', '.join(query_columns)
+
+        code = (
+            "PRAGMA foreign_keys=off;\n"
+            f"CREATE TABLE {table_name} ({query_columns});\n"
+            "PRAGMA foreign_keys=on;"
+        )
+        self._run_sql_code(pipeline_database, code)
+        return
```

Comment:
This part of the code has a lot in common with similarly named methods in `SQLiteDataConverter` that I'm thinking we should try to resolve. (One of the most important software engineering maxims is "DRY — Don't repeat yourself"). I'm noticing that none of these methods really make use of the calling instance (`self`) e.g. through member variables. This means that they could be converted to common utility methods, e.g. in an `src/graphnet/data/sql.py` module, that could contain all general functions that are relevant to SQL(ite)?

Reply:
I have moved most of the sqlite stuff to `data/utils.py`. I have kept` _create_table` because it creates the table scheme and I feel having this here makes sense. Making a general purpose function for this is outside of the scope of what I'd like to spend my time on right now. So if we could perhaps let it be, I'd appreciate it.